docs/rfcs/public-api-get-editor-for-audit-log.md

```diff
@@ -0,0 +1,238 @@
+then apply a mechanism to get the correct editor in api layer when making calls to the Web gRPC service.
+
+# Solutions
+
+## Problem 1: save the maintainer of API key when we create API key in Console
+
+We can add a column `maintainer` in api_key table so the maintainer can be saved. For old api keys, we can use creator
+in audit logs by filter type = 400 (create API key event) in audit_log table, select editor and `entity_id` 
+⇒ migration data for old `api_key` records and migration data for old `audig_log` records.
+
+Also, we need to hide the api key in the audit log snapshot, by adding a column `api_key` as new field to save api key
+and the old `id` column would be the UUID of the API key.
+
+<div style="text-align: center">
+  <img width="13%" src="/docs/rfcs/images/api-key-table-add-maintainer.png" alt="api_key table ERD" style="padding-right: 10px">
+  <img width="35%" src="/docs/rfcs/images/audit-log-snapshot.png" alt="audit log snapshot example">
+</div>
+
+**The message request of CreateAPIKey Web api should be:**
+
+```protobuf
+message CreateAPIKeyRequest {
+  string environment_namespace = 1;
+  string name = 2;
+  account.APIKey.Role role = 3;
+  string maintainer = 4;
+}
+```
+
+As we add new properties to `api_key` object, the response of the `GetAPIKeyBySearchingAllEnvironments` API 
+and `ListAPIKeys` API also return an extra `maintainer` and `id` field in `api_key` object:
+
+```protobuf
+message EnvironmentAPIKey {
+  string environment_namespace = 1 [deprecated = true];
+  APIKey api_key = 2;
+  bool environment_disabled = 3;
+  string project_id = 4 [deprecated = true];
+  environment.EnvironmentV2 environment = 5;
+  string project_url_code = 6;
+  string maintainer = 7;
+  string id = 8;
+}
+```
+
+## Problem 2: Get the correct editor in API layer when making call to Web service
+As we have saved API Key unique id, we can show it in the audit log like this (or just embed it in the api key name):
+```
+<key_name>-<id>-<creator_email> Feature flag has been created
+```
+I suggest 2 options:
+  - Let the maintainer be the creator of the API key.
+  - We let the client decide the editor of the action by adding editor email in the request body of public API.
+
+### Option 1: Let the editor be the creator of the API key <span style="color:green">(preferred)</span>
+
+We can extract the APIKey maintainer in function `getEnvironmentAPIKey` as we will update the response of 
+`GetAPIKeyBySearchingAllEnvironments` (https://github.com/bucketeer-io/bucketeer/blob/main/pkg/account/api/api_key.go#L455)
+
+Now to save the API key maintainer to audit log, we need to overwrite the editor, 
+the idea is to use context to share information between services.
+
+The `maintainer` value will be added into context before forward to Web gRPC service:
+
+```go
+const APIKeyMaintainerMDKey string = "apikey-maintainer"
+const APIKeyNameMDKey string = "apikey-name"
+const APIKeyIDMDKey string = "apikey-id"
```

Comment:
We should pass the API Key instead so we can use it in the audit log.

For example.
![Screenshot 2024-11-21 at 11 19 51 AM](https://github.com/user-attachments/assets/b1962b95-19c7-4378-95a4-8878b8236f58)

Reply:
Oh you mean audit log subscriber can query db and extract data from that key? Yeah that make sense

Reply:
Understood!